src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -92,6 +92,22 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(initEntryAssembly));
             }
 
+            // Initialize COM apartment
+            MethodDesc initApartmentState = startup.GetMethod("InitializeApartmentState", null);
+            if (initApartmentState != null)
+            {
+                if (_mainMethod.HasCustomAttribute("System", "STAThreadAttribute"))
+                {
+                    codeStream.EmitLdc(0 /* ApartmentState.STA */);
+                    codeStream.Emit(ILOpcode.call, emitter.NewToken(initApartmentState));
+                }
+                if (_mainMethod.HasCustomAttribute("System", "MTAThreadAttribute"))
```

Comment:
This structure is conveniently going to throw exception at runtime if one gets both STAThread and MTAThread attributes specified.